File: backend/templates/python/src/run/run_manager.py
import asyncio
import json
import sys
import time

from run_context import ctx
from influx.influx_service import InfluxService
from stress_test import StressTest
from run.user_runner import UserRunner


class RunManager:
    audit_exchange = None

    def __init__(self, test: StressTest):
        self.users = {}
        self.should_stop = False
        self.test = test
        self.influx = InfluxService()

    async def message_handler(self, message):
        data = message
        print("Client Received message", data["type"])
        if data["type"] == "startUser":
            self.start_user(data["userId"])
        elif data["type"] == "stopUser":
            await self.stop_user(data["userId"])
        elif data["type"] == "stopAllUsers":
            await self.stop_all_users()
            return False
        else:
            print("Client Unknown message type", data["type"], file=sys.stderr)
        return True

    def start_user(self, user_id: str):
        if user_id in self.users:
            print(f"User {user_id} is already running")
            return
        user_runner = UserRunner(user_id, self.test)
        self.users[user_id] = user_runner
        user_runner.start()

    async def stop_user(self, user_id: str):
        print(f"Stopping user {user_id}")
        if user_id not in self.users:
            print(f"User {user_id} is not running")
            return
        await self.users[user_id].stop()

    async def stop_all_users(self):
        print("Stopping all users")
        user_tasks = []
        for user_id in self.users:
            user_tasks.append(self.stop_user(user_id))
        self.should_stop = True
        await asyncio.gather(*user_tasks)

    async def run(self):
        try:
            while not self.should_stop or not self.all_users_stopped():
                await asyncio.sleep(1)
                await self.influx.write(
                    'running_users',
                    {
                        'value': float(len(self.users)),
                    },
                    {
                        'testId': ctx.test_id,
                        'runId': ctx.run_id,
                    },
                )
            print(
                f"All users stopped: {self.should_stop}, {self.all_users_stopped()}")
            print("Client Runner finished", flush=True)
        except Exception as e:
            print("Run manager main run thread crashed", e, flush=True)

    def all_users_stopped(self):
        for user_id in self.users:
            if self.users[user_id].status != 'stopped':
                return False
        return True

File: backend/templates/typescript/src/run/RunManager.ts
import { influxProxy } from '../influx/influxProxy.js';
import { ctx } from '../run.context.js';
import { StressTest } from '../StressTest.js';
import { UserRunner } from './UserRunner.js';

export class RunManager {
  private readonly users: Record<string, UserRunner> = {};
  private shouldStop = false;

  constructor(private readonly test: StressTest) {}

  public startUser(userId: string): void {
    if (this.users[userId]) {
      console.error(`User ${userId} is already running`);
      return;
    }
    this.users[userId] = new UserRunner(userId, this.test);
    this.users[userId].start();
  }

  public stopUser(userId: string): void {
    console.log(`Stopping user ${userId}`);
    if (!this.users[userId]) {
      console.error(`User ${userId} is not running`);
      return;
    }
    this.users[userId].stop();
  }

  public stopAllUsers(): void {
    console.log('Stopping all users');
    for (const userId in this.users) {
      this.stopUser(userId);
    }
    this.shouldStop = true;
  }

  public async run(): Promise<void> {
    while (!this.shouldStop || !this.AllUsersStopped()) {
      await new Promise((resolve) => setTimeout(resolve, 1000));
      await influxProxy.write({
        measurement: 'running_users',
        fields: {
          value: Object.keys(this.users).length,
        },
        tags: {
          testId: ctx.testId,
          runId: ctx.runId,
        },
      });
    }
    console.log('All users stopped', this.shouldStop, this.AllUsersStopped());
  }

  private AllUsersStopped(): boolean {
    for (const userId in this.users) {
      if (this.users[userId].status !== 'stopped') {
        return false;
      }
    }
    return true;
  }
}
